packages/@ember/debug/lib/warn.ts

```diff
@@ -9,7 +9,11 @@ export interface WarnOptions {
 }
 
 export type RegisterHandlerFunc = (handler: HandlerCallback<WarnOptions>) => void;
-export type WarnFunc = (message: string, test?: boolean, options?: WarnOptions) => void;
+export type WarnFunc = (
+  message: string,
+  test: boolean | WarnOptions,
+  options?: WarnOptions
+) => void;
```

Comment:
While this will mostly work, the right way to do this looks a little bit different, because we want to avoid it being legal for someone to pass the options *twice*, which this signature would allow. Instead, we want to use an [overload](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads). You can see how that would work, using an interface type instead, [here][p]. Also, the actual implementation *requires* that the options object be passed (it will assert if it's not).

I believe this should do the trick:

```suggestion
export interface WarnFunc {
  (message: string, options: WarnOptions): void;
  (
    message: string,
    test: boolean,
    options: WarnOptions
  ): void;
}
```

[p]: https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgOpyiA8gBzMAexAGdkBvAKGWWABMAuZYsKUAcwG4KBfCi0SLEQp0mAGIBXEAnJVkACgC2EYsThsIjZqxBsANMgJ5CJRqOzGixAJSMAbgTpdqSlWo1aW7A5GaMARgQEADYQcCAGRvhWZhgW0SS2yA5OPHy0EAjBGCihYMgA7nGx4lIIXBQA9JXIAAJgxAC0EAAeOJlgzVBQBFAURZjyAOQEABbIkOMgBCpD1lzVdQ3NbR1dPX0DIMNjExBTM8RDPlASEPP9cTvjk8jTsydnBmQ0DMhDoxDBwQRDyNwXLbXPYHB7kV6MD5fH5-AELGr1JqtdoITrQDaXQYjG77O6HY7guiQz7fX7-Z4Q94kmH-C4UIA